### PR TITLE
CORE-7854 Backchain resolution first version

### PR DESCRIPTION
**Pull request by lankydan**
Rough initial implementation. Raising this PR since I'm going on PTO for a week and people might want to have a look.

The main code to look at regarding backchain resolution is `TransactionBackchainResolutionFlow` and the other classes in its package.

Persistence code has been hacked together to get an E2E manual test working. This will be primarily replaced with @mbrkic-r3's work.

Temporary APIs were added to support my testing as the finality code doesn't exist yet.

### CODE REVIEW

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    val transactionIds = LinkedHashSet<SecureHash>()
+    private val nonDupeHash = HashMap<SecureHash, SecureHash>()
+    private fun dedupe(sh: SecureHash): SecureHash = nonDupeHash.getOrPut(sh) { sh }
+
+    /**
+     * Add a transaction to the to-be-sorted set of transactions.
+     * @param txId The ID of the transaction.
+     * @param dependentIds the IDs of all the transactions [txId] depends on.
+     */
+    fun add(txIdp: SecureHash, dependentIds: Set<SecureHash>) {
```
Comment (MrMatthewLayton) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt]: Not keen on the naming of `txIdp` here. I don't know what `p` is, and it should be `transactionId`.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    /**
+     * Return the sorted list of transaction IDs.
+     */
+    fun complete(): ArrayList<SecureHash> {
+        val visited = HashSet<SecureHash>(transactionIds.size)
+        val result = ArrayList<SecureHash>(transactionIds.size)
+
+        fun visit(txId: SecureHash) {
```
Comment (MrMatthewLayton) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt]: `transactionId`

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        if (resolvingTransactionId != other.resolvingTransactionId) return false
+        if (originalTransactionsToRetrieve != other.originalTransactionsToRetrieve) return false
+        if (session != other.session) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = resolvingTransactionId.hashCode()
```
Comment (MrMatthewLayton) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt:127]: Naive implementation of an [FNV hash](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function). Java provides an API to do this for you.

```kotlin
import java.util.Objects

override fun hashCode(): Int {
    return Objects.hash(resolvingTransactionId, originalTransactionsToRetrieve, session)
}
```

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+        if (sortedTransactionIds.isNotEmpty()) {
+            session.send(TransactionBackchainRequest.Stop)
+        }
+
+        return sortedTransactionIds
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
```
Comment (MrMatthewLayton) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt:114]: Whilst this is only a style preference, you can reduce this to a single expression:

```kotlin
override fun equals(other: Any?): Boolean {
    return this === other 
        || other is TransactionBackchainReceiverFlow
        && other.resolvingTransactionId == resolvingTransactionId
        && other.originalTransactionsToRetrieve == originalTransactionsToRetrieve
        && other.session == session
}
```

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/AbstractUtxoLedgerExternalEventFactory.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import net.corda.data.persistence.EntityResponse
+import net.corda.flow.external.events.factory.ExternalEventFactory
+import net.corda.flow.external.events.factory.ExternalEventRecord
+import net.corda.flow.state.FlowCheckpoint
+import net.corda.schema.Schemas
+import net.corda.virtualnode.toAvro
+import java.nio.ByteBuffer
+import java.time.Clock
+
+abstract class AbstractUtxoLedgerExternalEventFactory<PARAMETERS : Any>(
```
Comment (MrMatthewLayton) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/AbstractUtxoLedgerExternalEventFactory.kt]: To align with Java/Kotlin naming conventions for generics, I'd maybe consider renaming this to `P` or even just `T`.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoSignedTransactionImpl.kt [hunk 6]
```diff
@@ -66,4 +52,23 @@ data class UtxoSignedTransactionImpl(
     override fun toLedgerTransaction(): UtxoLedgerTransaction {
         return UtxoLedgerTransactionImpl(wireTransaction, serializationService)
     }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
```
Comment (MrMatthewLayton) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoSignedTransactionImpl.kt:118]: Again, could be reduced to a single expression, but not massively fussed either way.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt (hunk 1 above)
Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt]: Sure, I copied and pasted this from C4 code base and I would have wrote the same thing if someone else did this :)

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt (hunk 3 above)
Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt:127]: This was the hashcode that intellij generated. Its only needed for some assertions in my unit tests.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt (hunk 4 above)
Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainReceiverFlow.kt:114]: I had intellij generate this for me.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/AbstractUtxoLedgerExternalEventFactory.kt (hunk 5 above)
Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/external/events/AbstractUtxoLedgerExternalEventFactory.kt]: `PARAMTERS` is used because `ExternalEventFactory` has 3 generic types where one of them is named `PARAMETERS`.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistTransactionRequestHandler.kt [hunk 7]
```diff
@@ -26,15 +29,17 @@ class UtxoPersistTransactionRequestHandler(
 
     override fun execute(): List<Record<*, *>> {
         // Feed all produced and consumed states through any token observers defined for them
-        val producedTokens = transaction.getProducedStates().toTokens(tokenObservers)
-        val consumedTokens = transaction.getConsumedStates().toTokens(tokenObservers)
+//        val producedTokens = transaction.getProducedStates().toTokens(tokenObservers)
+//        val consumedTokens = transaction.getConsumedStates().toTokens(tokenObservers)
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistTransactionRequestHandler.kt]: What's with the commented out code here?

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 8]
```diff
@@ -18,12 +19,6 @@ class UtxoPersistenceServiceImpl constructor(
     private val utcClock: Clock
 ) : UtxoPersistenceService {
 
-    override fun findTransaction(id: String): SignedTransactionContainer? {
-        return sandbox.getEntityManagerFactory().createEntityManager().transaction { em ->
-            repository.findTransaction(em, id)
-        }
-    }
-
     override fun persistTransaction(transaction: UtxoTransactionReader) {
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: Nitpick - blank line

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt [hunk 9]
```diff
@@ -80,4 +75,89 @@ class UtxoPersistenceServiceImpl constructor(
...
+    }
+
+    override fun findTransaction(id: String, transactionStatus: String): SignedTransactionContainer? {
+        return sandbox.getEntityManagerFactory().createEntityManager().transaction { em ->
+            val status = repository.findTransactionStatus(em, id)
+            if (status == transactionStatus) {
+                repository.findTransaction(em, id)
+            } else {
+                null
+            }
```
Comment (blsemo) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/utxo/impl/UtxoPersistenceServiceImpl.kt]: Shouldn't that be atomic, i.e. use some clever SQL statement?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+                        .flatMap { it.dependencies }
+                        .map {
+                            // should send a Payload.Failure back for the non-existent transaction
+                            // a dependency not existing locally is a FATAL error, something seriously wrong has occurred.
+                            utxoLedgerPersistenceService.find(it)
+                                ?: throw CordaRuntimeException("Requested transaction does not exist locally")
+                        }
+                        .associateBy(UtxoSignedTransaction::id)
+                        .toMap(predictedTransactions)
+
```
Comment (blsemo) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt]: I don't quite understand what the benefit of this predicted transaction loading is - we still load them one by one, we might never need them (as the peer might already know them from an overlapping dependency tree).

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt [hunk 11]
```diff
@@ -25,9 +26,8 @@ data class UtxoTransactionBuilderImpl(
     override val attachments: List<SecureHash> = emptyList(),
     override val commands: List<Command> = emptyList(),
     private val signatories: Set<PublicKey> = emptySet(),
-    override val inputStateAndRefs: List<StateAndRef<*>> = emptyList(),
+    override val inputStateRefs: List<StateRef> = emptyList(),
     override val referenceInputStateAndRefs: List<StateAndRef<*>> = emptyList(),
```
Comment (blsemo) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt]: Why did you change input state and refs to StateRefs, but not references states?

**Review by blsemo — COMMENTED** [review]
Looks generally good, I've got a few questions/comments.
Also, the demo app looks a bit fishy...

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceService.kt [hunk 12]
```diff
@@ -34,5 +56,9 @@ interface UtxoLedgerPersistenceService {
      * @throws CordaPersistenceException if an error happens during find operation.
      */
     @Suspendable
-    fun find(id: SecureHash): UtxoSignedTransaction?
+    fun find(id: SecureHash, transactionStatus: TransactionStatus = VERIFIED): UtxoSignedTransaction?
 }
+
+enum class TransactionExistenceStatus {
```
Comment (MrMatthewLayton) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceService.kt]: Nitpicking - can this be moved into its own file so that file names and type match?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt (hunk 11 above)
Comment (MrMatthewLayton) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/transaction/UtxoTransactionBuilderImpl.kt]: @blsemo I think @vlajos might have done this in a separate PR

File: testing/cpbs/ledger-utxo-demo-app/src/main/kotlin/net/cordapp/demo/utxo/UtxoDemoFlow.kt [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+                    /   \
+                tx2       \
+                /           \         tx9
+            tx1             tx6 \    /   \
+               \          /       tx8     tx11
+                tx3 --- /--- tx7 /  \    /
+               \      /              tx10
+                tx4 /
+
+             */
```
Comment (MrMatthewLayton) [inline testing/cpbs/ledger-utxo-demo-app/src/main/kotlin/net/cordapp/demo/utxo/UtxoDemoFlow.kt]: LOVE the detail in this comment!

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt (hunk 10 above)
Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt]: The benefit isn't so great here because we're fetching transactions in batches of 1, but on larger batches it gives the responding flow a chance to pre-fetch transactions before the resolving flow requests them. 

This _should_ provide a performance improvement since both sides (the responding and resolving flows) can execute code at the same time. The resolving node will be persisting the received transactions while the responding flow pre-fetches transactions. 

When the responding flow does get told what flows the resolving node wants to retrieve it likely has already retrieved them and removes the need to then go to the db worker at that point.

It is true that we might not need some of the transactions that we've already pre-fetched, so it results in wasted work in this scenario. So I guess there would always be some wasted work with this solution if the resolving node already has any of the transactions in the backchain, if it has none then it will always use the pre-fetched transactions until issuance is reached.

We could remove this logic if we want upon thinking about it. In terms of performance saving, its not the largest, while pre-fetching on the responding flow the resolving flow is just persisting transactions. Furthermore, it might reduce the latency of a single flow the effect on the throughput of a node running many flows would be negligible to possibly negative due to the wasted work.

Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt]: Dicussed with @blsemo, we will leave it in for now and remove it in a separate PR. This means we'll have the git history here so adding it back in will take less effort if we decide to do so in the future.

**Review by MrMatthewLayton — DISMISSED** [review]

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    private val nonDupeHash = HashMap<SecureHash, SecureHash>()
+    private fun dedupe(sh: SecureHash): SecureHash = nonDupeHash.getOrPut(sh) { sh }
+
+    /**
+     * Add a transaction to the to-be-sorted set of transactions.
+     * @param transactionId The ID of the transaction.
+     * @param dependentIds the IDs of all the transactions [transactionId] depends on.
+     */
+    fun add(transactionId: SecureHash, dependentIds: Set<SecureHash>) {
+        val dedupedTransactionId = dedupe(transactionId)
```
Comment (mbrkic-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt:22]: Doesn't this always return the same value that was provided?

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        require(transactionIds.add(dedupedTransactionId)) { "Transaction ID $dedupedTransactionId already seen" }
+        dependentIds.forEach {
```
Comment (mbrkic-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt:24]: I think this might be simplified by using provided values without any transformations and depth-first search

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * Return the sorted list of transaction IDs.
+     */
+    fun complete(): ArrayList<SecureHash> {
+        val visited = HashSet<SecureHash>(transactionIds.size)
+        val result = ArrayList<SecureHash>(transactionIds.size)
+
+        fun visit(transactionId: SecureHash) {
+            if (visited.add(transactionId)) {
+                forwardGraph[transactionId]?.forEach(::visit)
```
Comment (mbrkic-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt:40]: Not sure what is the expected number of transactions that has to be supported, whether recursion should be replaced, i.e. to use stack explicitly so code doesn't depend on (default) JVM stack size

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt (hunk 14 above)
Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TopologicalSort.kt:22]: That's what I think as well. I copied this from our C4 code so didn't want to change it.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    @Suspendable
+    override fun call() {
+        // might want to send transaction ids from resolving flow that were in the dependencies but already verified so that they can be
+        // removed from the predicted collection of transactions
+        val predictedTransactions = linkedMapOf<SecureHash, UtxoSignedTransaction>()
+        while (true) {
+            when (val request = session.receive<TransactionBackchainRequest>()) {
+                is TransactionBackchainRequest.Get -> {
+                    val transactions = request.transactionIds.map { id ->
+                        predictedTransactions[id]?.also { predictedTransactions.remove(id) }
```
Comment (mbrkic-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt:27]: Can be reduced to `predictedTransactions.remove(id)`

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainVerifierImpl.kt [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    // that can still be done by sending long backchains
+
+    // return a boolean or just throw an exception?
+    @Suspendable
+    override fun verify(resolvingTransactionId: SecureHash, topologicalSort: TopologicalSort): Boolean {
+        val sortedTransactions = topologicalSort.complete().iterator()
+
+        for (transactionId in sortedTransactions) {
+            // Consider moving to separate private method as it might reduce the fiber stack when coming back into the loop
+            val transaction = utxoLedgerPersistenceService.find(transactionId, UNVERIFIED)
```
Comment (mbrkic-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainVerifierImpl.kt:36]: As there will be 1:1 mapping between transaction and status, (existing) find by transaction ID can be used and status checked after it's retrieved. That also allows detecting that transaction was already verified (not sure if such scenario could happen though).

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainResolutionFlow.kt [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+                    originalTransactionsToRetrieve,
+                    session
+                )
+            )
+            log.debug {
+                "Backchain resolution of ${transaction.id} - Retrieved dependencies of $originalTransactionsToRetrieve from its " +
+                        "backchain, beginning verification before storing the transactions locally"
+            }
+            try {
+                transactionBackchainVerifier.verify(transaction.id, topologicalSort)
```
Comment (mbrkic-r3) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainResolutionFlow.kt]: This returns Boolean that should be checked

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt (hunk 17 above)
Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/backchain/TransactionBackchainSenderFlow.kt:27]: I'm taking this code out in a later pr so will leave as is.

File: processors/flow-processor/build.gradle [hunk 20]
```diff
@@ -41,6 +41,8 @@ dependencies {
     runtimeOnly project(':components:ledger:ledger-consensual-flow')
     runtimeOnly project(':components:ledger:ledger-utxo-flow')
     runtimeOnly project(':libs:ledger:ledger-consensual-data')
+    runtimeOnly project(':components:ledger:ledger-utxo-flow')
```
Comment (vlajos) [inline processors/flow-processor/build.gradle:44]: I think this one is already there 2 lines above.

Comment (lankydan) [reply processors/flow-processor/build.gradle:44]: I'll address in a separate PR.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceService.kt [hunk 21]
```diff
@@ -15,6 +15,7 @@ interface UtxoLedgerPersistenceService {
     /**
      * Persist a [UtxoSignedTransaction] to the store.
      *
+     * @param transaction Consensual signed transaction to persist.
```
Comment (vlajos) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerPersistenceService.kt:18]: Consensual?

**Review by vlajos — APPROVED** [review]
It looks essentially OK. But not an easy reading... :)
And I do not see anything which could not be tackled in a follow up PR.

**Review by ronanbrowne — APPROVED** [review]
LGTM (build)
